libcst/metadata/type_inference_provider.py

```diff
@@ -0,0 +1,38 @@
+from typing import Optional
+
+import libcst as cst
+from libcst import BatchableMetadataProvider
+from libcst.metadata import PositionProvider
+
+
+class TypeInferenceProvider(BatchableMetadataProvider[str]):
+    METADATA_DEPENDENCIES = (PositionProvider,)
+    is_cache_required = True
+
+    def __init__(self, cache: object):
+        super().__init__(cache)
+        lookup = {}
+        for item in cache:
+            location = item["location"]
+            start = location["start"]
+            end = location["stop"]
+            lookup[(start["line"], start["column"], end["line"], end["column"])] = item[
```

Comment:
It'd be interesting to run this over IGSRV and log if we ever get duplicate ranges. That was one concern I had with this approach. We probably won't, but if it did happen, we might want to start considering node types in addition to positional ranges.

Reply:
Make sense, I can run some test and get some statistic numbers.

Reply:
Did you ever run this on IGSRV?

Reply:
We want to finish end to end integration first and then run on IGSRV to collect more examples in test suite. Back to when we just started the integration work, Rey run on IGSRV files to find mismatch positions and report to Pyre team (Shannon). Shannon already fixed many reported issues. So TypeInference provider assumes that the position always matches, if it doesn't match, then no inferred type is returned.
Once we landed this in IGSRV, I'll run on more IGSRV files to collect issues to report and build test suites.